lib/hsa/hsaRfiMaskOutput.cpp

```diff
@@ -66,40 +72,72 @@ hsa_signal_t hsaRfiMaskOutput::execute(int gpu_frame_id, hsa_signal_t precede_si
 void hsaRfiMaskOutput::finalize_frame(int frame_id) {
     hsaCommand::finalize_frame(frame_id);
 
-    if (get_rfi_zeroed(_network_buf, _network_buf_id)) {
-        uint8_t* frame_mask = _rfi_mask_output_buf->frames[_rfi_mask_output_buf_id];
-        uint32_t total_lost = 0;
-        // Copy RFI mask to array
-        for (uint32_t subframe = 0; subframe < _num_sub_frames; ++subframe) {
-            uint32_t lost_in_subframe = 0;
-            for (uint32_t i = subframe * _sub_frame_samples / _sk_step;
-                 i < (subframe + 1) * _sub_frame_samples / _sk_step; i++) {
-                assert(i < (uint32_t)_rfi_mask_output_buf->frame_size);
-                if (frame_mask[i] == 1)
-                    lost_in_subframe += _sk_step;
-            }
-            atomic_add_lost_timesamples(_output_buf, _output_buf_id, lost_in_subframe);
-            mark_frame_full(_output_buf, unique_name.c_str(), _output_buf_id);
-            _output_buf_id = (_output_buf_id + 1) % _output_buf->num_frames;
+    uint8_t* mask_frame = _rfi_mask_output_buf->frames[_rfi_mask_output_buf_id];
+    uint8_t* lost_samples_frame = _lost_samples_buf->frames[_lost_samples_buf_id];
+
+    // Add the number of flagged samples to the metadata
+    uint32_t total_flagged = 0;
+    uint32_t total_net_lost_samples = 0;
+    uint32_t sub_frame_mask_len = _sub_frame_samples / _sk_step;
+
+    for (uint32_t subframe = 0; subframe < _num_sub_frames; ++subframe) {
+
+        // Total number of samples flaged as RFI
+        uint32_t flagged_samples = 0;
+
+        // The number of samples flaged as RFI less the number of
+        // samples flaged as missing (i.e. packet loss/currupt packet).
+        uint32_t net_lost_samples = 0;
 
-            total_lost += lost_in_subframe;
+        for (uint32_t i = 0; i < sub_frame_mask_len; ++i) {
+            uint32_t mask_idx = subframe * sub_frame_mask_len + i;
+            assert(mask_idx < (uint32_t)_rfi_mask_output_buf->frame_size);
+            if (mask_frame[mask_idx] == 1) {
+                flagged_samples += _sk_step;
+                net_lost_samples += _sk_step;
+                // Remove any samples which were also counted as lost samples
+                // in this block of data.
```

Comment:
Were did they get double counted?

Reply:
They get counted in `zeroSamples.cpp`.  However zeroed packets often look like RFI and trigger the RFI zeroing, which means if the RFI system adds to the lost sample count because of zeroing it needs to correct for the zeroing of the packet, but only if the two are coincident in time.

Reply:
Ok, makes sense now!